### ISSUE TITLE
Would it be a good idea to have multiple zim files share the same favicon URL when possible?

### ISSUE THREAD

**Issue by automactic** [kiwix/libkiwix#609]
There are a lot of zim files that share (seemingly) the same favicon, yet they have different favicon urls. I can surely retrieving them individually, but that seems like a waste. Especially when you consider a more extreme case: there are around 400 TED zim files, it makes less sense to make 400 GET requests when one will do.

What if we have multiple zim files share the same favicon url if their favicon look the same? 

<img src="https://user-images.githubusercontent.com/8294252/130865765-b2577a81-fdb0-4ab2-8689-9cd90d90d0c1.png" width="300">

**Comment by mgautierfr** [comment]
One solution would be to serve the favicon under a path base on a checksum of the favicon's content. This way, identical favicon would have the same (unique) path.

To do so, we should extend the library.xml to store the checksum (along with the actual data) and the opds stream to point to a different path, independent of the zim file.
The server/library would need to map the checksum to the content (when reading the library.xml) and "serve" the correct content corrsesponding to the checksum.

This is a question we have to keep in mind when we answer https://github.com/kiwix/libkiwix/issues/625 and https://github.com/kiwix/libkiwix/issues/631, even if we don't implement it now.

**Comment by kelson42** [comment]
@rgaudin Would that be possible to do that cleanly in the cache or reverse proxy for library.kiwix.org?

**Comment by rgaudin** [comment]
@kelson42 I don't understand what you're asking. The problem being the client making multiple requests, I don't see how any server-side configuration only would save requests from the client…

**Comment by kelson42** [comment]
@rgaudin I agree with @mgautierfr. The question is where to handle this smart identification of same illustrations:
* In library.xml
* in kiwix-serve
* in library.kiwix.org cache/cdn infrastructure

After thinking twice about it, I believe like @mgautierfr that if we handle it in library.kiwix.org this would be better. Actually, I would consider to implement  https://github.com/kiwix/overview/issues/59 first. This could help as well for https://github.com/kiwix/k8s/issues/65, the CMS would just do the job and all the rest would more or less follow automatically.

**Comment by rgaudin** [comment]
I also think we'd need to tackle the multiple illustrations thing first ; including the devicePixelRatio stuff.

> in library.kiwix.org cache/cdn infrastructure

I don't think this would be help much. We'd still make tons of requests the first time and because we want them to be fixable via CMS, it would mean a short ttl (so many etag-check requests) or additional purge code on cms.
Something based on redirects would help but we'd still have the original-URIs requests and I suppose it would be sub-optimal as I imagine those requests are concurrent. Also this would be custom code for this.
Doesn't seem worth the effort.

> in kiwix-serve

definitely where it makes the most sense as it's the one responsible for reading and serving. Probably needs to be done anyway in case it's not in the source XML (unless this becomes mandatory).

> In library.xml

that would be an optimization but given all of what we're discussing here is related to the special use case of the Kiwix demo and for which we have control over the catalog, it'd be both necessary and easy.

**Comment by mgautierfr** [comment]
If we want (and we probably want it) that kiwix-serve serve the illustration content using a url base of the content itself (and not its name or book's name), we want that kiwix-serve to know the content based url of the illustration it have to serve.
- It can be computed by kiwix-serve itself (as it will have to serve the content itself, it can make a foosum on it)
- It can be provided to kiwix-serve.

Secondly, we need a way to tell the clients which contentbased url to use for each book. It can be done:
- By putting directly the content based url in the catalog, favicon url, ....
- Using a redirection from a book name based url to the content based url

The two points are orthogonal and so we have 4 possibilities:
- kiwix-serve compute the url and put it directly in the catalog : kiwix-serve will have to open the illustration for all book advertised in the catalog. It can be pretty time consuming at initialisation or library reloading. But nothing have to be done in CMS and we avoid the redirect request.
- kiwix-serve compute the url and use redirection : Nothing have to be done in CMS, kiwix-serve can trust the library.xml. When a request is made on the bookname based url, kiwix-serve open the illustration, computes its foosum, fill its internal redirection map and answer the request with a redirection to the contentbased url.
- kiwix-serve is given the url mapping and put url directly in the catalog: This a new kind of information to give to kiwix-serve. And it add a feature to kiwix-serve pretty close from a generic mapping `{this_url}=>{this item in this book}`. CMS has to do all the work of generating this mapping and creating a library.xml using the content based url (if we want kiwix-serve to trust it)
- kiwix-serve is given the url mapping and use redirection: Same kind of new feature. But as different books may shared the same illustration (and so the same contentbased url), we also need a reverse mapping `{illustration of this boo} => {this_url}`.

Saying that, I think the best is the second solution. At the cost of redirection requests.
